drivers/peci/peci_mchp_xec.c

```diff
@@ -223,27 +225,37 @@ static int peci_xec_read(const struct device *dev, struct peci_msg *msg)
 	int ret;
 	uint8_t tx_fcs;
 	uint8_t bytes_rcvd;
+	uint8_t wait_timeout_cnt;
 	struct peci_buf *rx_buf = &msg->rx_buffer;
 	PECI_Type *base = peci_xec_config.base;
 
 	/* Attempt to read data from RX FIFO */
 	bytes_rcvd = 0;
 	for (i = 0; i < (rx_buf->len + PECI_FCS_LEN); i++) {
-		/* Check if data available */
-		if (!(base->STATUS2 & MCHP_PECI_STS2_RFE)) {
-			if (i == 0) {
-				/* Get write block FCS just for debug */
-				tx_fcs = base->RD_DATA;
-				LOG_DBG("TX FCS %x", tx_fcs);
-			} else if (i == (rx_buf->len + 1)) {
-				/* Get read block FCS, but don't count it */
-				rx_buf->buf[i-1] = base->RD_DATA;
-			} else {
-				/* Get response */
-				rx_buf->buf[i-1] = base->RD_DATA;
-				bytes_rcvd++;
+		/* Worst case timeout will be 1msec (100 * 10usec) */
+		wait_timeout_cnt = PECI_RX_BUF_FILL_WAIT_RETRY;
+		/* Wait for read buffer to fill up */
+		while (base->STATUS2 & MCHP_PECI_STS2_RFE) {
+			k_usleep(PECI_IO_DELAY);
+			wait_timeout_cnt--;
+			if (!wait_timeout_cnt) {
+				LOG_WRN("Rx buffer empty");
+				return -ETIMEDOUT;
 			}
 		}
+
+		if (i == 0) {
+			/* Get write block FCS just for debug */
+			tx_fcs = base->RD_DATA;
+			LOG_DBG("TX FCS %x", tx_fcs);
+		} else if (i == (rx_buf->len + 1)) {
+			/* Get read block FCS, but don't count it */
+			rx_buf->buf[i-1] = base->RD_DATA;
```

Comment:
Isn't this writing past the end of the buffer?  Buffer length is peci_buf.len specified by the caller but the loop is iterating over peci_buf.len + 2 sinece PECI_FCS_LEN is 2.

Reply:
@scottwcpg this is something I document in the header that rx buffer always need to allocate an extra byte in rx_buffer.
 * Note: Frame check sequence byte is added into rx buffer, need to allocate
 * an additional byte for this in rx buffer.
struct peci_buf {
	uint8_t *buf;  (payload + fcs)
	size_t len;
};
(as opposed to add an explicit fcs field)
struct peci_buf {
	uint8_t *buf;  (payload)
        uint8_t fcs;
	size_t len;
};

If you think this may lead to problems, we can revisit the PECI driver API